### PR TITLE
Witgen for multiplicities in LogUp in PIL

### PR DESCRIPTION
**Pull request by onurinanc**
Opened this PR to discuss on the change files related to the issue #1573

```
columns: {"main.m_logup_multiplicity": [Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([2, 0, 0, 0]) }, Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([2, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }]}
Vec<T> parts: [[Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([2, 0, 0, 0]) }, Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([1, 0, 0, 0]) }, Bn254Field { value: BigInt([2, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }]]
String parts: ["main.m_logup_multiplicity"]
columns.len(): 1
witness_cols: [("main.y", [Bn254Field { value: BigInt([2, 0, 0, 0]) }, Bn254Field { value: BigInt([6, 0, 0, 0]) }, Bn254Field { value: BigInt([3, 0, 0, 0]) }, Bn254Field { value: BigInt([7, 0, 0, 0]) }, Bn254Field { value: BigInt([5, 0, 0, 0]) }, Bn254Field { value: BigInt([3, 0, 0, 0]) }, Bn254Field { value: BigInt([7, 0, 0, 0]) }, Bn254Field { value: BigInt([4, 0, 0, 0]) }]), ("main.m_logup_multiplicity", [Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }, Bn254Field { value: BigInt([0, 0, 0, 0]) }])]
```

Using take_witness_col_values inside the FixedLookup machine, we have the correct values of the `main.m_logup_multiplicitiy` column, However, when it comes to witness_cols, the values are changed.
The above issue is fixed.

With this PR, we have a witness generation for multiplicities in LogUp / bus argument.

### CODE REVIEW

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs]: Okay, so I deleted the namespace from the struct definition, and got the namespace as a function

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (Schaeff) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:206]: This seems to me like it still has the same issue: `self.fixed_data.witness_cols` contains all witness columns in the pil, across all namespaces. So what this function does is rather `get_first_namespace` which I doubt would work if we had for example two machines with each a `col witness m_logup_multiplicity`. The same goes for the `has_logup_multiplicity_column`.

I would suggest adding a pipeline test which does witgen for two machines with each a multiplicity column.

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:206]: You're right. For example, should it work for both "main.m_logup_multiplicity" and "arith.m_logup_multiplicity"? Or, should we collect all the namespaces and return None from try_new by adding the following as same as the double_sorted_witness_machine?

`if namespaces.len() > 1 {
            // columns are not in the same namespace, fail
            return None;
        }`

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:206]: Also, the function `take_witness_col_values()` in fixed_lookup_machine should return both `main.m_logup_multiplicity` and `arith.m_logup_multiplicity` right?

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:206]: What do you think about adding `if namespaces.len() > 1 { // columns are not in the same namespace, fail return None; }` mechanism in this PR, and adding an example with 2 different namespaces together with a witgen fix in the next PR?

Comment (georgwiese) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:206]: Yeah, since the `FixedLookup` currently is a strange machine in that it is responsible for columns of different namespaces, it would be responsible for all multiplicity columns associated with any lookup in `connecting_identities`. We could store a `multiplicity_columns: BTreeMap<u64, PolyID>` (identity ID -> multiplicity poly ID).

Figuring out this mapping is a bit of a hassle and obsolete after #1378 anyway though, so I think it could be fair to defer to a different PR. But I think failing in the presence of > 1 namespace is overly strict, we could fail if there is > 1 column called `*.m_logup_multiplicity`, right?

Maybe instead of `has_logup_multiplicity_column`, you can store `logup_multiplicity_column: Option<PolyID>`. Then, instead of calling `self.namespaced(MULTIPLICITY_LOOKUP_COLUMN)`, you can do `logup_multiplicity_column_name = logup_multiplicity_column.map(|poly_id| self.fixed_data.column_name(poly_id))`.

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (georgwiese) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:245]:
```suggestion
            // No fixed column, so this FixedLookup instance does nothing
            .unwrap_or(0) as u64;
```

This wasn't obvious to me at first glance, think a comment helps :)

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (georgwiese) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:445]:
```suggestion
        let mut witness_col_values = HashMap::new();
        if self.has_logup_multiplicity_column {
            assert!(self.multiplicities.len() <= 1, "LogUp witness generation not yet supported for > 1 lookups");
            log::trace!("Detected LogUp Multiplicity Column");
            for multiplicity in std::mem::take(&mut self.multiplicities).into_values() {
                let mut values = vec![];
                for row in 0..self.degree as usize {
                    values.push(T::from(multiplicity.get(&row).cloned().unwrap_or_default()));
                }
                witness_col_values.insert(self.namespaced(MULTIPLICITY_LOOKUP_COLUMN), values);
            }
        }
        witness_col_values
```

How about this? I don't think there is a need to build a Map with a big number of 0s first.

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:445]: Yes, this looks better

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (georgwiese) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:230]:
```suggestion
        // This currently just takes one element with the correct name
```

**Review by georgwiese — COMMENTED** [review]
Cool, think this is good for a first version. We should should implement #1378 soon, so that we can get rid of the pattern matching and support more than one lookup.

I think the test needs to be adjusted after #1802 is merged (currently in queue), and after that can actually be tested with Goldilocks as well and no longer needs to be blacklisted in the parsing test.

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:311]: This also stores the value if we don't have/need a multiplicities column, right?

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (georgwiese) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:232]: Yes, but figuring out that mapping is not straight-forward (would need to find and analyze the polynomial identities belonging to the lookup), whereas it's trivial after #1378 (because we'll have an explicit annotation). So, in my opinion, we should get that done first.

File: pipeline/tests/powdr_std.rs
Comment (chriseth) [inline pipeline/tests/powdr_std.rs:460]: Is this correct? Doesn't it automatically use the extension field now?

File: executor/src/witgen/mod.rs
Comment (chriseth) [inline executor/src/witgen/mod.rs:265]: I would say either `get columns from machines` or `get the columns from the machines`

File: executor/src/witgen/machines/machine_extractor.rs
Comment (chriseth) [inline executor/src/witgen/machines/machine_extractor.rs:178]: At this point, it's irrelevant that it's a multiplicity column, so I think a better name for ths function is `witness_columns`

Comment (chriseth) [reply executor/src/witgen/machines/machine_extractor.rs:178]: Also, are you sure this works like this? What about the identities that involve the multiplicity columns? Doesn't the machine extractor create a machine for that? Or shouldn't we at least also remove those identities?

Comment (chriseth) [reply executor/src/witgen/machines/machine_extractor.rs:178]: Ah, maybe they are all filtered out because they involver higher-stage witness columns. Can you check nevertheless?

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:194]: Why not return `self.logup_multiplicitiy_columns.iter().cloned().collect()`?

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:416]: Hm, do we really want to merge it like this? Why not create `self.multiplicities` as a vector of size `self.degree` from the start? It would be so much faster, I think.

Comment (chriseth) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:416]: And it should already be a `T` - or do you think one conversion from integer to field element is cheaper than incrementing a field element for each access?

File: executor/src/witgen/machines/machine_extractor.rs
Comment (onurinanc) [reply executor/src/witgen/machines/machine_extractor.rs]: I'm going to check this

File: pipeline/tests/powdr_std.rs
Comment (onurinanc) [reply pipeline/tests/powdr_std.rs]: Removed it

File: executor/src/witgen/machines/machine_extractor.rs
Comment (onurinanc) [reply executor/src/witgen/machines/machine_extractor.rs]: > Also, are you sure this works like this? What about the identities that involve the multiplicity columns? Doesn't the machine extractor create a machine for that? Or shouldn't we at least also remove those identities?

How can I get them to remove?

Comment (onurinanc) [reply executor/src/witgen/machines/machine_extractor.rs]: > Ah, maybe they are all filtered out because they involver higher-stage witness columns. Can you check nevertheless?

Yes, we're filtering here: https://github.com/powdr-labs/powdr/blob/main/executor/src/witgen/mod.rs#L220

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs]: Converting is more expensive, you're right. Now, fixed it

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:230]:
```suggestion
            .find(|col| SymbolPath::from_str(&col.poly.name).name() == MULTIPLICITY_LOOKUP_COLUMN)
```

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:314]: In the end, we create a full vector of the full column and store it in memory. Doesn't it make sense to start with a vector (instead of a hashmap) to begin with?

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs]: But, then I need to change this part

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:183]: I already forgot what the key is, can you add it?

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:306]: Please only do this if we have a multiplicities column.

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:183]: The key is `identity_id`. Do you want me to add as a comment?

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs:183]: Okay

File: executor/src/witgen/machines/fixed_lookup_machine.rs
Comment (chriseth) [inline executor/src/witgen/machines/fixed_lookup_machine.rs:307]:
```suggestion
        if self.logup_multiplicity_column.is_some() {
```

Comment (onurinanc) [reply executor/src/witgen/machines/fixed_lookup_machine.rs]: changed

**Comment by chriseth** [comment]
I think there is some kind of merge problem here.

**Comment by onurinanc** [comment]
> I think there is some kind of merge problem here.

resolved

**Comment by onurinanc** [comment]
> > I think there is some kind of merge problem here.
> 
> resolved

@chriseth?

**Review by chriseth — APPROVED** [review]

**Pull request merged**
